cam6_4_023: SCAM-SE feature addition plus bugfixes and some refactoring.

jtruesdal opened the pull request:
This update includes some refactoring of SCAM, a few bugfixes, and adding the capability to use spectral elements dycore to do vertical transport in the column.  The SE feature addition follows the E3SM implementation where a complete coarse resolution (ne3np4) of the SE dycore is initialized but only a single element is run through vertical transport with all element subcolumns being copies of the single column chosen by scmlat, scmlon.  

Like the Eulerian version, SCAM-SE also has a bit for bit test to validate an exact run through the same physics as the full 3d model.  Because SCAM updates the solution using a slightly different order of operations, the bfb capability is tested by making a special diagnostic run of CAM where the 3d model derives the phys/dyn tendency each time step and then recalculates the prognostic solution using the derived tendencies and SCAM's prognostic equation. This new solution (which is less precise (roundoff) due to the change in order of operations) is substituted for the full 3d solution at each time step of the model run.  The substitution of the roundoff state in the 3d run allows SCAM to reproduce (BFB) each time step using the captured tendencies in the cam iop history file. 

The SCAM-SE vertical advection skips the horizontal step and derives the floating level tendency based on the IOP prescribed vertical velocity. The floating levels are subsequently remapped at the end of the vertically Lagrangian dynamics step.

This PR also includes an update to .gitmodules for cice and CDEPS.
	- update cice to fix scam regression failure
	- update cdeps to fix CDEPS regression test build failures


Closes #957 
Closes #853 
Closes #742

jedwards4b commented on src/dynamics/se/apply_iop_forcing.F90:
```diff
@@ -0,0 +1,172 @@
+module apply_iop_forcing_mod
+
+use shr_kind_mod,   only:r8 => shr_kind_r8, i8 => shr_kind_i8
+use pmgrid
```
Please add only clause.

jtruesdal replied:
Done.

jedwards4b reviewed (changes requested):
Hi John - thank you for your work on this, it's clear that you've put a lot of time and effort into it.  
There are a lot of changes using user mods or test mods that should be done in config_compsets.xml - I know there are several ways to do this - user mods are the worst in my opinion, please consider changing.    I also strongly feel that this PR should include a change to [shr_scam_mod.F90 ](https://github.com/ESCOMP/CESM_share/blob/main/src/shr_scam_mod.F90#L717) to remove the requirement for MCT.

jtruesdal commented on Externals.cfg:
```diff
@@ -45,7 +45,7 @@ required = True
 tag = share1.0.17
 protocol = git
-repo_url = https://github.com/ESCOMP/CESM_share
+repo_url = https://github.com/jtruesdal/CESM_share
```
i'll double check this and get rid of it.  Thanks.

jtruesdal replied:
Tag name should be share1.0.17_scamdev.  I had a few mods for shr_scam_mod that turned up in testing for this PR. If you can help with the mods for getting rid of MCT I'm happy to include them, test and create the PR to CESM_share.  In the meantime I'll fix the tag reference above.

jtruesdal replied:
@jedwards4b forgot to 'at' you on the previous question.  Repeating here:

>Tag name should be share1.0.17_scamdev. I had a few mods for shr_scam_mod that turned up in testing for this PR. If you can help with the mods for getting rid of MCT I'm happy to include them, test and create the PR to CESM_share. In the meantime I'll fix the tag reference above.

jedwards4b replied:
Sorry I haven't looked into it yet, I'll try to do it tomorrow.

jedwards4b commented:
@cacraigucar I have to disagree,  I think that user mods really obfuscate things - but I have a meeting with John at 2 today to look at this and maybe what we should do is create another branch so that we can compare and contrast the two approaches.

cacraigucar commented:
> @cacraigucar I have to disagree, I think that user mods really obfuscate things - but I have a meeting with John at 2 today to look at this and maybe what we should do is create another branch so that we can compare and contrast the two approaches.

I'm getting ready to start reviewing this PR and was curious about any proposed changes that might be made due to your meeting yesterday?   @jedwards4b @jtruesdal

cacraigucar commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1251 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+      call get_start_count(ncid, varid, scmlat, scmlon, ioptimeidx, strt4, cnt4)
+      call wrap_get_vara_realx (ncid,varid,strt4,cnt4,f12vmrobs)
+   else
+      write(6,*)'using column value of f12vmr from boundary data as global volume mixing ratio'
```
This is obviously really old code.  Need to change `write(6` to use `iulog`.  This pertains throughout this routine

jtruesdal replied:
Fixed.

cacraigucar commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1251 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+   endif
+
+   status = nf90_inq_varid( ncid, 'Tg', varid   )
+   if (status .ne. nf90_noerr) then
```
Could you update the logical operators from `.ne.` to` /=` throughout this code?

jtruesdal replied:
Updated to use F90 relational operators throughout the code.

cacraigucar reviewed (changes requested):
Submitting the items I have found so far - have not completed the review by any means

jtruesdal commented on cime_config/usermods_dirs/scam_gateIII/shell_commands:
```diff
@@ -8,8 +8,17 @@
+./xmlchange STOP_OPTION=nhours
+./xmlchange STOP_N=479
+
+if [ `./xmlquery --value CAM_DYCORE` == 'se' ]; then
```
Agreed. The latest set of mods move this functionality from usermods to new compsets and related config routines.  I'm testing now and will ask for another review when its available.

jtruesdal replied:
See conversation below with Cheryl.

jtruesdal commented:
> > @cacraigucar` I have to disagree, I think that user mods really obfuscate things - but I have a meeting with John at 2 today to look at this and maybe what we should do is create another branch so that we can compare and contrast the two approaches.

> I'm getting ready to start reviewing this PR and was curious about any proposed changes that might be made due to your meeting yesterday? @jedwards4b @jtruesdal

We did meet and as it turns out I already had a mostly complete set of mods for moving to a compset version of the different IOPs. I wanted to see what the effect of pushing the IOP configuration under cime_config would be. So I created a new branch with them to compare to the current way of doing things. 

https://github.com/jtruesdal/CAM-1/compare/b567e51..jtruesdal:CAM-1:scam_dev_exp_w_compsets?expand=1

The mods add a cam6 configuration option for each IOP.  I also added an alias for each IOP although we could keep one main IOP alias and allow users to switch between the various IOPS by specifying the correct longname.  Having the IOP configure options also allow the user to run any IOP case using any of the other configuration options available in the longname.  We no longer need the usermods scam directories and they have been deleted in the branch.  Although if someone wants to add their own IOP then they will most likely do so through the usermods directory, so from the user perspective nothing has changed in that respect.

Whether they should be a configured via compsets is still an open question.  Although they are not configuring for new physics parameterizations, one of the changes being added in this commit is the ability to modify the CLUBB parameterization to be more consistent with the type of forcing represented by the IOP.  After looking at the comparison, they don't seem to overly complicate the xml files and these options make it easier to use the framework default configurations for each IOP via the compset longnames instead of having to individually specify them via usermods. I agree with Jim that this is a better way to setup the standard IOP's while still allowing usermods for user defined IOPs or user_nl_cam to tweak the setup like we do with other compsets.

jtruesdal commented:
@jedwards4b wrote:
> I also strongly feel that this PR should include a change to [shr_scam_mod.F90 ](https://github.com/ESCOMP/CESM_share/blob/main/src/shr_scam_mod.F90#L717)to remove the requirement for MCT.

I'm happy to include this change but after looking at the code in question I don't know how to make the proper changes for NUOPC.  Would you be able to help or provide a set of mods to use?  I'm happy to test along with the rest of the PR.

jtruesdal commented on Externals.cfg:
```diff
@@ -13,24 +13,24 @@ local_path = components/cice5
 
 [cdeps]
-tag = cdeps1.0.24
+branch = cdeps1.0.24_scamdev
```
I will change these to the official branch names as the PR's for each of these branches goes through.  I think most should be timely enough but land may take longer than we want to wait, in which case I can do without and keep the coldstart option for land.  I will add to a misc PR or create another small one to include the land update when it is approved.

cacraigucar replied:
I tried to find the .gitmodules file and I am not finding it via this PR.  Did you forget to "git add" it when you removed the Externals.cfg?

cacraigucar replied:
I agree that all the externals you needed to have updated have independently been done in CAM already, so changes for externals are no longer expected.

jtruesdal commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -314,8 +314,9 @@
 <bnd_topo hgrid="ne120np4" npg="2">atm/cam/topo/se/ne120pg2_nc3000_Co015_Fi001_PF_nullRR_Nsw010_20171012.nc</bnd_topo>
 <bnd_topo hgrid="ne240np4" npg="2">atm/cam/topo/se/ne240pg2_nc3000_Co008_Fi001_PF_nullRR_Nsw005_20171014.nc</bnd_topo>
-
+<bnd_topo hgrid="ne3np4"          >atm/cam/topo/se/ne3np4_gmted2010_modis_bedmachine_nc0540_Laplace1000_noleak_20230717.nc</bnd_topo>
```
It has the appropriate metadata and has been imported.

jtruesdal commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -314,8 +314,9 @@
-
+<bnd_topo hgrid="ne3np4"          >atm/cam/topo/se/ne3np4_gmted2010_modis_bedmachine_nc0540_Laplace1000_noleak_20230717.nc</bnd_topo>
 <bnd_topo hgrid="ne3np4"   npg="3">atm/cam/topo/se/ne3pg3_gmted2010_modis_bedmachine_nc0540_Laplace1000_noleak_20230209.nc</bnd_topo>
+<bnd_topo hgrid="ne4np4"          >atm/cam/topo/USGS-gtopo30_ne4np4_16x.c20160612.nc</bnd_topo>
```
This file is not needed and I've removed it.

jtruesdal commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -1895,7 +1896,9 @@
 
 <!-- Dry Dep surface data file needed by prognostic MAM on unstructured grid only. -->
-<drydep_srf_file hgrid="ne3np4"  npg="3" >atm/cam/chem/trop_mam/atmsrf_ne3np4.pg3_c221214.nc</drydep_srf_file>
+<drydep_srf_file hgrid="ne3np4"          >atm/cam/chem/trop_mam/atmsrf_ne3np4_230718.nc</drydep_srf_file>
+<drydep_srf_file hgrid="ne3np4"  npg="3" >atm/cam/chem/trop_mam/atmsrf_ne3np4.pg3_221214.nc</drydep_srf_file>
+<drydep_srf_file hgrid="ne4np4"          >atm/cam/chem/trop_mam/atmsrf_ne4np4_from_0.23x0.31_181018.nc</drydep_srf_file>
```
Added the metadata to the ne3 datasets and imported. np4 is not needed and was removed.

jtruesdal commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -2857,12 +2860,16 @@
+<scale_dry_air_mass                               scam="1"  >   0.0D0    </scale_dry_air_mass>
+<scm_zadv_q  dyn="eul"                            scam="1"  >   slt      </scm_zadv_q>
+<!-- default namelist for SE SCAM mode -->
+<ncdata  dyn="se1"   hgrid="ne3np4"                scam="1"  >atm/cam/inic/se/CESM2.F2000climo.ne3np4.cam.i.0003-06-01-00000.nc</ncdata>
```
This file is one of 12 monthly files needed to initialize SCAM-SE.  I did make sure each has proper metadata and was imported into the repository.  The defaults file will also be updated once we make a decision on the best way to initialize the different IOPS (usermods or compset).

jtruesdal commented on src/control/history_scam.F90:
```diff
@@ -1,106 +1,219 @@
    end subroutine scm_intht
+!#######################################################################
+   subroutine initialize_iop_history()
+!
+! !DESCRIPTION:
+! !USES:
+    use constituents,     only: pcnst, cnst_name
+    use dycore,           only: dycore_is
+! !ARGUMENTS:
+    implicit none
+!
+! !CALLED FROM:
+!
+! !REVISION HISTORY:
+!
+!EOP
+!
+! !LOCAL VARIABLES:
```
Added description and removed rest of header.

jtruesdal commented on src/control/history_scam.F90:
```diff
@@ -1,106 +1,219 @@
+       dyngrid = 'unknown'
+    end if
+
+    if (trim(dyngrid) == 'gauss_grid') then
+    call addfld ('CLAT1&IC',  horiz_only,  'I', ' ','cos lat for bfb testing', gridname=trim(dyngrid))
+    call add_default ('CLAT1&IC',0,'I')
```
Done.  Thanks.

jtruesdal commented on src/control/ncdio_atm.F90:
```diff
@@ -93,7 +91,7 @@ subroutine infld_real_1d_2d(varname, ncid, dimname1,                        &
     type(io_desc_t), pointer  :: iodesc
     integer                   :: grid_id   ! grid ID for data mapping
-    integer                   :: i, j      ! indices
+    integer                   :: j         ! indice
```
fixed

jtruesdal commented on src/control/ncdio_atm.F90:
```diff
@@ -515,14 +494,11 @@ subroutine infld_real_2d_3d(varname, ncid, dimname1, dimname2,              &
     type(io_desc_t), pointer  :: iodesc
     integer                   :: grid_id   ! grid ID for data mapping
-    integer                   :: i, j, k   ! indices
+    integer                   :: j         ! indice
```
Fixed.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1251 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+!     Written by J.  Truesdale    August, 1996, revised January, 1998
+!
+!-----------------------------------------------------------------------
+!jt fix this circular depend	use phys_grid,           only: clat_p
+!jt	use commap,              only: latdeg, clat
```
Not an issue and removed these comments.

cacraigucar commented on src/dynamics/se/se_single_column_mod.F90:
```diff
@@ -0,0 +1,345 @@
+  tl_f = timelevel%n0
+  call TimeLevel_Qdp(timelevel, qsplit, tl_fqdp)
+
+  if (.not. use_camiop .and. get_nstep() .eq. 0) then
```
Please use `==` instead of `.eq.`

jtruesdal replied:
.done.

cacraigucar commented on src/dynamics/se/se_single_column_mod.F90:
```diff
@@ -0,0 +1,345 @@
+
+1000 continue
+
+    if (get_nstep() .le. 1) then
```
Please use `<=` instead of `.le.`

jtruesdal replied:
done

cacraigucar commented on src/dynamics/se/se_single_column_mod.F90:
```diff
@@ -0,0 +1,345 @@
+       qobs(:)=elem(ie_scm)%state%qdp(i_scm,j_scm,:,1,tl_fqdp)/elem(ie_scm)%state%dp3d(i_scm,j_scm,:,tl_f)
+    endif
+
+    if (get_nstep() .eq. 0) then
```
Another `.eq.` to update

jtruesdal replied:
done

cacraigucar commented on src/dynamics/se/se_single_column_mod.F90:
```diff
@@ -0,0 +1,345 @@
+      integer :: ierr
+#ifdef SPMD
+
+      call mpi_bcast(have_ps,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(have_tg,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(have_lhflx,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(have_shflx,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(have_t,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(have_q,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(have_u,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(have_v,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(have_omega,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(have_cldliq,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(have_divt,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(have_divq,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(have_divt3d,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(have_divq3d,1,mpi_logical,mstrid,mpicom,ierr)
+      call mpi_bcast(use_3dfrc,1,mpi_logical,mstrid,mpicom,ierr)
+
+      call mpi_bcast(psobs,1,mpi_real8,mstrid,mpicom,ierr)
+      call mpi_bcast(tground,1,mpi_real8,mstrid,mpicom,ierr)
+      call mpi_bcast(lhflxobs,1,mpi_real8,mstrid,mpicom,ierr)
+      call mpi_bcast(shflxobs,1,mpi_real8,mstrid,mpicom,ierr)
+
+      call mpi_bcast(tobs,nlev,mpi_real8,mstrid,mpicom,ierr)
+      call mpi_bcast(qobs,nlev,mpi_real8,mstrid,mpicom,ierr)
+      call mpi_bcast(uobs,nlev,mpi_real8,mstrid,mpicom,ierr)
+      call mpi_bcast(vobs,nlev,mpi_real8,mstrid,mpicom,ierr)
+      call mpi_bcast(cldliqobs,nlev,mpi_real8,mstrid,mpicom,ierr)
+      call mpi_bcast(wfld,nlev,mpi_real8,mstrid,mpicom,ierr)
+
+      call mpi_bcast(divt,nlev,mpi_real8,mstrid,mpicom,ierr)
+      call mpi_bcast(divq,nlev,mpi_real8,mstrid,mpicom,ierr)
+      call mpi_bcast(divt3d,nlev,mpi_real8,mstrid,mpicom,ierr)
+      call mpi_bcast(divq3d,nlev,mpi_real8,mstrid,mpicom,ierr)
```
The ierr should probably be checked after each of these mpi_bcast calls and the run halting if an error is detected

cacraigucar commented on src/dynamics/se/se_single_column_mod.F90:
```diff
@@ -0,0 +1,345 @@
+            do i=1, np
+               testlat=elem(ie)%spherep(i,j)%lat * rad2deg
+               testlon=elem(ie)%spherep(i,j)%lon * rad2deg
+               if (testlon .lt. 0._r8) testlon=testlon+360._r8
+               testval=abs(scmlat-testlat)+abs(scmposlon-testlon)
+               if (testval .lt. minpoint) then
```
A couple more `.lt.` which should be replaced with `<`

jtruesdal replied:
@cacraigucar I'll go through all the routines I touched and make the changes to relational operators.  Is it OK to change those lines which were not part of the original set of mods so as to leave the routine in an updated consistent state?  I will be making a short test as each file is changed to make sure that these mods don't introduce any additional changes.

jtruesdal replied:
done

cacraigucar commented on src/dynamics/eul/iop.F90:
```diff
@@ -130,1026 +100,37 @@ subroutine init_iop_fields()
+   if (present(ps)) ps(1,1,timelevel)           = psobs
+   if (present(t3)) t3(1,ioptop:,1,timelevel)   = tobs(ioptop:)
+   if (present(q3)) q3(1,ioptop:,1,1,timelevel) = qobs(ioptop:)
+!!!!!jt revert next line only for bfb
+   ioptop=1
```
Is this still valid and if so, should it become a namelist?

jtruesdal replied:
The repository SCAM code does not extrapolate forcing data outside the IOP pressure boundaries, filling those regions with 0's.  I've modified the code to fill these 0 regions with the background state from CAM.  I believe this is the proper behavior and I will remove this flag before committing the code.  For now it allows me to verify BFB results with the previous implementation.

jtruesdal replied:
Thanks.  I will go through these comments before tagging and make sure its removed.  I do appreciate the extra set of eyes on this.

cacraigucar replied:
This too will need to be removed before I can give my final approval

cacraigucar replied:
I believe this still needs to be removed

jedwards4b commented:
@jtruesdal looking at this this morning, I don't see any tests of scam on the ne grid listed in the [cam/cime_config/testdefs/testlist_cam.xml](https://github.com/jtruesdal/CAM-1/blob/scam_dev_exp/cime_config/testdefs/testlist_cam.xml)  Can you please add at least one test and also tell me what compset and resolution I should try.

jedwards4b commented:
I tried test SMS.ne3pg3_ne3pg3_mg37.FSCAM.derecho_intel and ran into missing input data: 
```
Model cam missing file drydep_srf_file = '/glade/campaign/cesm/cesmdata/inputdata/atm/cam/chem/trop_mam/atmsrf_ne3np4.pg3_221214.nc'
Trying to download file: 'atm/cam/chem/trop_mam/atmsrf_ne3np4.pg3_221214.nc' to path '/glade/campaign/cesm/cesmdata/inputdata/atm/cam/chem/trop_mam/atmsrf_ne3np4.pg3_221214.nc' using SVN protocol.
svn export failed with output:  and errput svn: E170000: URL 'https://svn-ccsm-inputdata.cgd.ucar.edu/trunk/inputdata/atm/cam/chem/trop_mam/atmsrf_ne3np4.pg3_221214.nc' doesn't exist


```

jedwards4b commented:
@jtruesdal I tried to run test SCT_D_Ln7.T42_T42_mg17.QPC6.derecho_intel.cam-scm_prep_c6 and SCT_D_P36_Ln7.T42_T42_mg17.QPC6.derecho_intel.cam-scm_prep_c6 and both fail with an mpi error.   Can you point me to an SCT test that runs on derecho - or izulme.

jedwards4b commented:
I also tried these tests on izumi and both fail.  I'm looking into the issue but where are the baselines and what tests have you run?

jtruesdal commented:
Thanks Jim.  You can hold for now, I see my SCT test failed also and am looking into it.

jtruesdal commented on src/physics/cam/check_energy.F90:
```diff
@@ -587,18 +590,38 @@ subroutine check_energy_fix(state, ptend, nstep, eshflx)
     ! disable the energy fix for offline driver
     heat_glob = 0._r8
 #endif
+! add (-) global mean total energy difference as heating
```
fixed.

jtruesdal replied:
Reworked the if statement and updated the comments.

jtruesdal commented on src/physics/cam/check_energy.F90:
```diff
@@ -587,18 +590,38 @@ subroutine check_energy_fix(state, ptend, nstep, eshflx)
+      heat_glob = 0._r8
+    endif
 ! add (-) global mean total energy difference as heating
     ptend%s(:ncol,:pver) = heat_glob
 
+    if (nstep > 0 .and. write_camiop) then
+      heat_out(:ncol) = heat_glob
+      call outfld('heat_glob',  heat_out(:ncol), pcols, lchnk)
+    endif
```
heat_glob is a private module variable and is set before this routine is called.  If this was a problem it would have showed up in the current regression testing as heat_glob is used elsewhere in this routine.

jtruesdal replied:
Regression testing aside I did some tracing and can't see how it could be unassigned.  Do see something I didn't catch?

jtruesdal commented on src/infrastructure/phys_grid.F90:
```diff
@@ -235,11 +238,39 @@ subroutine phys_grid_init()
+            minpoint=1000.0_r8
+            do i=1,columns_on_task
+               testpoint=abs(pos_dynlons(i)-pos_scmlon)+abs(dynlats(i)-scmlat)
+               if (testpoint .lt. minpoint) then
```
Done.

jtruesdal commented:
Jim I merged and tested your share changes on Izumi and Derecho. I also corrected the SCAM testing and added new regression tests for SE SCAM.  The code is now up to cam6_3_147.

jtruesdal commented:
>@jtruesdal looking at this this morning, I don't see any tests of scam on the ne grid listed in the [cam/cime_config/testdefs/testlist_cam.xml](https://github.com/jtruesdal/CAM-1/blob/scam_dev_exp/cime_config/testdefs/testlist_cam.xml) Can you please add at least one test and also tell me what compset and resolution I should try.

The 2 new regression tests are:  
SCT_D_Ln7_Vnuopc.ne3_ne3_mg37.QPC5.derecho_intel.cam-scm_prep
SCT_D_Ln7_Vnuopc.ne3_ne3_mg37.QPC6.izumi_gnu.cam-scm_prep_c6

details:

    PASS SCT_D_Ln7_Vnuopc.ne3_ne3_mg37.QPC6.izumi_gnu.cam-scm_prep_c6 CREATE_NEWCASE
    PASS SCT_D_Ln7_Vnuopc.ne3_ne3_mg37.QPC6.izumi_gnu.cam-scm_prep_c6 XML
    PASS SCT_D_Ln7_Vnuopc.ne3_ne3_mg37.QPC6.izumi_gnu.cam-scm_prep_c6 SETUP
    PASS SCT_D_Ln7_Vnuopc.ne3_ne3_mg37.QPC6.izumi_gnu.cam-scm_prep_c6 SHAREDLIB_BUILD time=308
    PASS SCT_D_Ln7_Vnuopc.ne3_ne3_mg37.QPC6.izumi_gnu.cam-scm_prep_c6 MODEL_BUILD time=560
    PASS SCT_D_Ln7_Vnuopc.ne3_ne3_mg37.QPC6.izumi_gnu.cam-scm_prep_c6 SUBMIT
    PASS SCT_D_Ln7_Vnuopc.ne3_ne3_mg37.QPC6.izumi_gnu.cam-scm_prep_c6 RUN time=41
    PASS SCT_D_Ln7_Vnuopc.ne3_ne3_mg37.QPC6.izumi_gnu.cam-scm_prep_c6 COMPARE_base_default
    PASS SCT_D_Ln7_Vnuopc.ne3_ne3_mg37.QPC6.izumi_gnu.cam-scm_prep_c6 MEMLEAK insufficient data for memleak test
    PASS SCT_D_Ln7_Vnuopc.ne3_ne3_mg37.QPC6.izumi_gnu.cam-scm_prep_c6 SHORT_TERM_ARCHIVER

I will remove the CAM5 test once the PR is approved.

cacraigucar reviewed (changes requested):
These changes and the few unresolved ones are my final comments.

jtruesdal commented on src/dynamics/se/se_single_column_mod.F90:
```diff
@@ -0,0 +1,376 @@
+    do k=1, NLEV
+       if (tobs(k)  /=  0) then
+          thelev=k
+          go to 1000
```
Rewrote this original E3SM code minloc intrinsic.  Thanks for the catch.

jtruesdal commented on src/dynamics/se/se_single_column_mod.F90:
```diff
@@ -0,0 +1,376 @@
+
+    if (get_nstep()  <=  1) then
+       do k=1,thelev-1
+          tobs(k)=elem(ie_scm)%state%T(i_scm,j_scm,k,tl_f)
+          qobs(k)=elem(ie_scm)%state%qdp(i_scm,j_scm,k,1,tl_fqdp)/elem(ie_scm)%state%dp3d(i_scm,j_scm,k,tl_f)
```
Fixed and checking the rest of my changes for long lines.

jtruesdal commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -2873,12 +2879,130 @@
+        'TGCLDLWP','GCLDLWP'
+</fincl1>
+
+<ncdata             hgrid="ne3np4"    nlev="32"       scam_iop="rico"      >atm/cam/inic/se/CESM2.F2000climo.ne3np4.cam.i.0003-07-01-00000.nc</ncdata>
+<ncdata             hgrid="64x128"    nlev="32"       scam_iop="rico"      >atm/cam/inic/gaus/CESM2.F2000climo.64x128.cam.i.0003-07-01-00000.nc</ncdata>
+<iopfile                                              scam_iop="rico"      >atm/cam/scam/iop/RICO_3day_4scam.nc</iopfile>
+
+<ncdata             hgrid="ne3np4"    nlev="32"       scam_iop="SAS"       >atm/cam/inic/se/CESM2.F2000climo.ne3np4.cam.i.0003-06-01-00000.nc</ncdata>
+<ncdata             hgrid="64x128"    nlev="32"       scam_iop="SAS"       >atm/cam/inic/gaus/CESM2.F2000climo.64x128.cam.i.0003-06-01-00000.nc</ncdata>
```
The IOPs are field programs that took place during specific months of the year.  The initial data files which specify background state should be consistent with the IOP run dates.  I double checked all the dates to make sure they match the base date of the IOP files.

jtruesdal commented:
Added the drydep file to the repo via rimport.

nusbaume commented on src/dynamics/se/se_single_column_mod.F90:
```diff
@@ -0,0 +1,372 @@
+    u_in(:nlev) = elem(ie_scm)%state%v(i_scm,j_scm,1,:nlev,tl_f)
+    v_in(:nlev) = elem(ie_scm)%state%v(i_scm,j_scm,2,:nlev,tl_f)
+
+    t_phys_frc(:)   = elem(ie_scm)%derived%fT(i_scm,j_scm,:)
```
Just double-checking that `t_phys_frc` doesn't need to be divided by `dt` as well?

jtruesdal replied:
It is correct.  t_phys_frc is derived from phys_tend$dtdt returned from physics and is already a tendency whereas the quantity carried in derived%fQ is the change in state due to the physics and not already a tendency.

nusbaume commented on src/dynamics/se/se_single_column_mod.F90:
```diff
@@ -0,0 +1,372 @@
+    subroutine iop_broadcast()
+
+      !---------------------------------------------------------
+      ! Purpose: When running DP-CRM, broadcast relevant logical
```
Should this say SCAM?

```suggestion
      ! Purpose: When running SCAM, broadcast relevant logical
```

jtruesdal replied:
Yes. Changed as recommended.

nusbaume commented on src/dynamics/se/se_single_column_mod.F90:
```diff
@@ -0,0 +1,372 @@
+!=========================================================================
+    subroutine scm_dyn_grid_indicies(elem,scmlat,scmlon,ie_scm,i_scm,j_scm,indx_scm)
+
+      !---------------------------------------------------------
+      ! Purpose: When running DP-CRM, broadcast relevant logical
+      !   flags and data to all processors
+      !----------------------------------------------------------
```
I don't think this comment block is correct (?).

jtruesdal replied:
Corrected the comment.

nusbaume commented on src/dynamics/se/se_single_column_mod.F90:
```diff
@@ -0,0 +1,372 @@
+      enddo
+
+      if (ie_scm == 0 .or. i_scm == 0 .or. j_scm == 0 .or. indx_scm == 0) then
+         call endrun('Could not find closest SCM point on input datafile')
```
I might include the subroutine name in the `endrun` message here so that it is easier to find the source code if needed.

jtruesdal replied:
Thanks.  I added the subroutine name

nusbaume reviewed (changes requested):
Great work!  I realize it looks like there are a ton of change requests, but most of them should either be trivial, or are just general questions.  Thanks!

jtruesdal commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -2873,12 +2879,130 @@
+<scm_use_obs_uv                  scam="1"                   > .true.    </scm_use_obs_uv>
+<scale_dry_air_mass              scam="1"        dyn="se"   > 0.0D0     </scale_dry_air_mass>
+<scm_relaxation                  scam="1"                   > .true.    </scm_relaxation>
+<scm_relax_fincl   phys="cam6"   scam="1"  >
+  'T', 'bc_a1', 'bc_a4', 'dst_a1', 'dst_a2', 'dst_a3', 'ncl_a1', 'ncl_a2', 'ncl_a3',
+  'num_a1', 'num_a2', 'num_a3', 'num_a4', 'pom_a1', 'pom_a4', 'so4_a1', 'so4_a2', 'so4_a3', 'soa_a1', 'soa_a2'
+</scm_relax_fincl>
+<scm_relax_fincl   phys="cam5"   scam="1"  >
+  'T', 'bc_a1', 'bc_a4', 'dst_a1', 'dst_a2', 'dst_a3', 'ncl_a1', 'ncl_a2', 'ncl_a3',
+  'num_a1', 'num_a2', 'num_a3', 'num_a4', 'pom_a1', 'pom_a4', 'so4_a1', 'so4_a2', 'so4_a3', 'soa_a1', 'soa_a2'
+</scm_relax_fincl>
+<scm_relax_fincl   phys="cam_dev"scam="1"  >
+  'T', 'bc_a1', 'bc_a4', 'dst_a1', 'dst_a2', 'dst_a3', 'ncl_a1', 'ncl_a2', 'ncl_a3',
+  'num_a1', 'num_a2', 'num_a3', 'num_a4', 'pom_a1', 'pom_a4', 'so4_a1', 'so4_a2', 'so4_a3', 'soa_a1', 'soa_a2'
```
These are OK to define for everything but CAM4.  I don't know how to exclude CAM4 other than only defining these defaults for CAM5,6,DEV.

jtruesdal commented on bld/config_files/definition.xml:
```diff
@@ -147,12 +147,16 @@ Turn on CO2 cycle in biogeochemistry model: 0 => no, 1 => yes.
 <entry id="scam" valid_values="0,1" value="0">
-Configure CAM for single column mode: 0=off, 1=on.  This option only
-supported for the Eulerian dycore.
+Configure CAM for single column mode and specify an IOP.
```
Done.

jtruesdal commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -219,6 +219,11 @@
 
 <ncdata dyn="se" hgrid="ne3np4"   nlev="32"             ic_ymd="101" >atm/cam/inic/se/cam6_QPC6_topo_ne3pg3_mg37_L32_01-01-31_c221214.nc</ncdata>
+<ncdata dyn="se" hgrid="ne3np4"   nlev="32"             ic_ymd="901" >atm/cam/inic/se/CESM2.F2000climo.ne3np4.cam.i.0003-09-01-00000.nc</ncdata>
+<ncdata dyn="se" hgrid="ne3np4"   nlev="30"             ic_ymd="101" >atm/cam/inic/se/cami_0000-01-01_ne3np4_L30_c120315.nc</ncdata>
+<ncdata dyn="se" hgrid="ne3np4"   nlev="30"             ic_ymd="901" >atm/cam/inic/se/cami_0000-01-01_ne3np4_L30_c120315.nc</ncdata>
+<ncdata dyn="se" hgrid="ne3np4"   nlev="26"             ic_ymd="101" >atm/cam/inic/se/cami_0000-01-01_ne3np4_L26_c120525.nc</ncdata>
+<ncdata dyn="se" hgrid="ne3np4"   nlev="26"             ic_ymd="901" >atm/cam/inic/se/cami_0000-01-01_ne3np4_L26_c120525.nc</ncdata>
```
The naming for the date is consistent with the date of the file that they were directly interpolated from not the date they were created.  The creation date in the metadata is correct though.  I realize the naming is inconsistent with the current guide lines but there is a method to the madness and I would ask that they be allowed to remain as is especially since these files have been committed to the data repository already.

jtruesdal commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -2873,12 +2879,130 @@
+<iopfile                                              scam_iop="arm97"     >atm/cam/scam/iop/ARM97_4scam.nc</iopfile>
+<co2vmr                                               scam_iop="arm97"     > 368.9e-6  </co2vmr>
+
+<ncdata             hgrid="ne3np4"    nlev="32"       scam_iop="atex"      >atm/cam/inic/se/CESM2.F2000climo.ne3np4.cam.i.0003-02-01-00000.nc</ncdata>
```
I've added those comments.  The information came from Andrew Gettleman's SCAM paper.

jtruesdal commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -2873,12 +2879,130 @@
+<!-- "IOP: CAM Forcing file" -->
+<ncdata            hgrid="ne3np4"    nlev="32"        scam_iop="camfrc"     >atm/cam/inic/se/CESM2.F2000climo.ne3np4.cam.i.0003-06-01-00000.nc</ncdata>
+<ncdata            hgrid="64x128"    nlev="32"        scam_iop="camfrc"     >atm/cam/inic/gaus/CESM2.F2000climo.64x128.cam.i.0003-06-01-00000.nc</ncdata>
+<iopfile                                              scam_iop="camfrc"     >atm/cam/scam/iop/ARM97_4scam.nc</iopfile>
```
Just putting some reasonable defaults in there that the user can modify via usermods_dir.  I put in a comment speaking to that point.

jtruesdal commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -5860,6 +5860,12 @@ Use the SCAM-IOP specified observed water vapor at each time step instead of for
+<entry id="scm_use_3dfrc" type="logical" category="scam"
+       group="scam_nl" valid_values="">
+Use the SCAM-IOP 3d forcing if true, use combination of dycore vertical advection and iop horiz advection if false
+Default:
```
Done.

jtruesdal commented on cime_config/SystemTests/sct.py:
```diff
@@ -47,25 +49,38 @@ def _case_two_setup(self):
+            append_to_user_nl_files(caseroot = self._get_caseroot(), component = "cam", contents = "scale_dry_air_mass     = 0.0D0")
+        else:
+            append_to_user_nl_files(caseroot = self._get_caseroot(), component = "cam", contents = "ncdata = '../"+case_name+".cam.i."+RUN_STARTDATE+"-00000.nc'")
+            self._case.set_value("CAM_CONFIG_OPTS","{} -scam camfrc ".format(CAM_CONFIG_OPTS))
```
Done.

jtruesdal commented on cime_config/config_component.xml:
```diff
@@ -178,6 +177,25 @@
+      <value compset="_CAM.*%SCAMRICO">-scam rico</value>
+      <value compset="_CAM.*%SCAMSPARTICUS">-scam sparticus</value>
+      <value compset="_CAM.*%SCAMTOGA2">-scam togaII</value>
+      <value compset="_CAM.*%SCAMTOGA2">-scam user</value>
```
Good catch.  Thanks.

jtruesdal commented on cime_config/config_compsets.xml:
```diff
@@ -121,8 +121,98 @@
+
+  <compset>
+    <alias>FSCAMCGILSS6</alias>
+    <lname>2000_CAM60%SCAMCgilss6_CLM50%SP_CICE%PRES_DOCN%DOM_SROF_SGLC_SWAV</lname>
```
I changed it to uppercase.

jtruesdal commented on cime_config/testdefs/testlist_cam.xml:
```diff
@@ -1151,6 +1151,14 @@
+    </machines>
+    <options>
+      <option name="wallclock">00:30:00</option>
+    </options>
```
Done.

jtruesdal commented on cime_config/testdefs/testlist_cam.xml:
```diff
@@ -1335,6 +1343,14 @@
+    </machines>
+    <options>
+      <option name="wallclock">00:30:00</option>
+    </options>
```
Done.

jtruesdal commented on cime_config/testdefs/testlist_cam.xml:
```diff
@@ -1446,7 +1462,7 @@
 
 <!--              CAM SCAM runs                      -->
-  <test compset="FSCAM" grid="T42_T42" name="SMS_D_Ln9_Vnuopc" testmods="cam/outfrq9s">
+  <test compset="FSCAMARM97" grid="T42_T42" name="SMS_D_Ln9_Vnuopc" testmods="cam/outfrq9s">
```
Done.

jtruesdal commented on cime_config/testdefs/testlist_cam.xml:
```diff
@@ -1455,7 +1471,7 @@
     </options>
   </test>
-  <test compset="FSCAM" grid="T42_T42" name="SMS_D_Ln9_Vnuopc" testmods="cam/scam_mpace_outfrq9s">
+  <test compset="FSCAMARM97" grid="T42_T42" name="SMS_D_Ln9_Vnuopc" testmods="cam/scam_mpace_outfrq9s">
```
Done.

jtruesdal commented on cime_config/usermods_dirs/scam_mandatory/shell_commands:
```diff
@@ -9,7 +9,7 @@
 # Note that clm cannot use initial conditions with SCAM -so will only use specified phenology
-# Only change if CLM_FORCE_COLDSTART exists.
-if [ `./xmlquery --value CLM_FORCE_COLDSTART |& grep -c 'ERROR'` -eq 0  ]; then
+# Only change if CLM_FORCE_COLDSTART exists and dycore is eularian
```
Changed.

jtruesdal commented on src/control/getinterpnetcdfdata.F90:
```diff
@@ -43,10 +42,13 @@ subroutine getinterpncdata( NCID, camlat, camlon, TimeIdx, &
    logical, intent(in)  :: have_surfdat  ! is surfdat provided
-   logical, intent(in)  :: fill_ends ! extrapolate the end values
+   logical, intent(in)  :: fill_ends     ! extrapolate the end values
+   logical, intent(in)  :: scm_crm_mode  ! extrapolate the end values
```
Corrected the comment.

jtruesdal commented on src/control/getinterpnetcdfdata.F90:
```diff
@@ -43,10 +42,13 @@ subroutine getinterpncdata( NCID, camlat, camlon, TimeIdx, &
    integer, intent(in)  :: npress        ! number of dataset pressure levels
    real(r8), intent(in) :: press(npress) ! dataset pressure levels
-   real(r8), intent(in) :: ps ! dataset pressure levels
+   real(r8), intent(in) :: ps            ! dataset pressure levels
```
Corrected.

jtruesdal commented on src/control/history_scam.F90:
```diff
@@ -1,106 +1,218 @@
+       call add_default ('CLAT1&IC',0,'I')
+       call addfld ('CLON1&IC',  horiz_only,  'I', ' ','cos lon for bfb testing', gridname=trim(outgrid))
+       call add_default ('CLON1&IC',0,'I')
+       call addfld ('PHI&IC',    horiz_only,  'I', ' ','lat for bfb testing', gridname=trim(outgrid))
+       call add_default ('PHI&IC',0,  'I')
+       call addfld ('LAM&IC',    horiz_only,  'I', ' ','lon for bfb testing', gridname=trim(outgrid))
```
Historical.  I'd have to look into what the ramifications of changing these would be.

jtruesdal commented on src/control/history_scam.F90:
```diff
@@ -1,106 +1,218 @@
+
+       call addfld ('fixmas',  horiz_only,   'A', 'percent','Mass fixer',gridname=trim(outgrid))
+       call add_default ('fixmas',2,' ')
+       call addfld ('beta',    horiz_only,   'A', 'percent','Mass fixer',gridname=trim(outgrid))
```
Beta is an energy fixer coefficient and the comment was updated

jtruesdal commented on src/control/history_scam.F90:
```diff
@@ -1,106 +1,218 @@
+    call add_default ('t',2,' ')
+    call addfld ('Tg',      horiz_only,   'A', 'K',      'Surface temperature (radiative) for scam',gridname='physgrid')
+    call add_default ('Tg',2,' ')
+    call addfld ('Ps',      horiz_only,   'A', 'Pa',     'Ps for scam',gridname=trim(outgrid))
```
Modified as recommended

jtruesdal commented on src/control/history_scam.F90:
```diff
@@ -1,106 +1,218 @@
 !
-      integer m,j        ! Indices
-      real(r8) dummy
+      character(len=100) outgrid
```
Yes. Changed

jtruesdal commented on src/control/history_scam.F90:
```diff
@@ -1,106 +1,218 @@
+
+! !LOCAL VARIABLES:
+    integer m
+    character(len=100) outgrid
```
Yes Changed.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -14,31 +14,46 @@ module scamMod
+public :: scam_readnl         ! read SCAM namelist options
+public :: readiopdata         ! read iop boundary data
+public :: setiopupdate        ! find index in iopboundary data for current time
+public :: plevs0              ! find index in iopboundary data for current time
```
Corrected.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -142,33 +167,34 @@ module scamMod
-logical, public ::  have_lhflx             = .false. ! dataset contains lhflx 
+logical, public ::  have_lhflx             = .false. ! dataset contains lhflx
 logical, public ::  have_shflx             = .false. ! dataset contains shflx
+logical, public ::  have_heat_glob         = .false. ! dataset contains shflx
```
Yes changed as recommended

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -179,41 +205,47 @@ module scamMod
+real(r8), public              ::  scm_relaxation_low      ! lowest level to apply relaxation
+real(r8), public              ::  scm_relaxation_high     ! highest level to apply relaxation
+real(r8), public              ::  scm_relax_top_p       = 1.e36_r8 ! upper bound for scm relaxation
+real(r8), public              ::  scm_relax_bot_p       = -1.e36_r8 !  lower bound for scm relaxation
```
Range for top of model should be 0. While bottom can be huge. Changed

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -179,41 +205,47 @@ module scamMod
+logical, public ::  scm_use_obs_uv         = .true. ! Use the SCAM-IOP observed u,v at each time step instead of forecasting.
 
-logical, public ::  scm_use_obs_qv         = .false. ! Use the SCAM-IOP specified observed qv  at each time step instead of forecasting.
+logical, public ::  scm_use_obs_qv         = .false. ! Use the SCAM-IOPobserved qv  at each time step instead of forecasting.
```
Changed to recommended comment

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+   implicit none
+#if ( defined RS6000 )
+   implicit automatic ( a-z )
+#endif
```
Nope

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+   if (status /= NF90_NOERR) then
+      status = nf90_inq_dimid (ncid, 'tsec', time_dimID )
+      if (status /= NF90_NOERR) then
+         if (masterproc) write(iulog,*) sub//':ERROR - readiopdata.F:Could not find dimension ID for time/tsec'
```
Implemented

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+      if (status /= NF90_NOERR) then
+         if (masterproc) write(iulog,*) sub//':ERROR - readiopdata.F:Could not find dimension ID for time/tsec'
+         status = NF90_CLOSE ( ncid )
+         call endrun
```
Added.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+      if (masterproc) write(iulog,*) sub//':ERROR - readiopdata.F:Could not find variable dim ID  for lev'
+      status = NF90_CLOSE ( ncid )
+      return
+   end if
```
Your correct and endrun is needed here.  There are only a few IOP variables that should result in an endrun when missing and dimension info is one of them.  Added endrun.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+   call handle_ncerr( nf90_inquire_dimension( ncid, lev_dimID, len=nlev ),&
+         'readiopdata.F90', __LINE__)
+
+   allocate(dplevs(nlev+1))
```
The handle_ncerr does check the status return of the nf90 call and prints it.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+      if (masterproc) write(iulog,*) sub//':ERROR - readiopdata.F:Could not find variable ID for lev'
+      status = NF90_CLOSE ( ncid )
+      return
+   end if
```
Thanks this should be an endrun.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+      if (masterproc) write(iulog,*) sub//':Could not find variable Ps'
+      if ( .not. scm_backfill_iop_w_init ) then
+         status = NF90_CLOSE( ncid )
+         return
```
I'll have to test these.  It does look like an error putting return in here as I agree we want the rest of the file to be processed.  All these variables are on the standard IOP which is maybe why it works correctly.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+   endif
+   if ( nlev == 1 ) then
+      if (masterproc) write(iulog,*) sub//':Error - Readiopdata.F: Ps too low!'
+      return
```
This is an error.  There must be a mismatch with the units between the lev dimension and Ps.  I've called endrun and updated the error message.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+      if (is_first_step()) write(iulog,*)'using column value of soltsi from boundary data as global solar tsi'
+   end if
+!=====================================================================
+!get global vmrs from camiop file
```
Yes. Changed

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+      if (masterproc) write(iulog,*) sub//':Could not find variable T'
+      if ( .not. scm_backfill_iop_w_init ) then
+         status = NF90_CLOSE( ncid )
+         return
```
Corrected logic and error messages.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+      if (masterproc) write(iulog,*) sub//':Could not find variable q'
+      if ( .not. scm_backfill_iop_w_init ) then
+         status = nf90_close( ncid )
+         return
```
Corrected logic and error messages.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+      if (masterproc) write(iulog,*) sub//':Could not find variable omega'
+      if ( .not. scm_backfill_iop_w_init ) then
+         status = nf90_close( ncid )
+         return
```
Corrected logic and error messages.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+   else
+      have_v = .true.
+   endif
+   call shr_sys_flush( iulog )
```
Not that I can tell.  This looks old and probably was needed at sometime in the distant past.  I removed these flushes.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+
+   deallocate(dplevs)
+
+   return
```
Deleted

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+  implicit none
+#if ( defined RS6000 )
+  implicit automatic (a-z)
+#endif
```
Nope. Deleted

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+   if ( ncdate > last_date .or. (ncdate == last_date &
+      .and. ncsec > last_sec))  then
+      if ( .not. scm_backfill_iop_w_init ) then
+         call endrun(sub//':ERROR - setiopupdate.c:Reached the end of the time varient dataset')
```
Changed as recommended.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+   if (masterproc) write(iulog,*) sub//':iop time index = ' , ioptimeidx
+#endif
+
+   return
```
Deleted

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+! $Id$
+! $Author$
+!
+!-----------------------------------------------------------------------
```
Deleted.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+!-----------------------------------------------------------------------
+
+!---------------------------Local workspace-----------------------------
+  integer k             ! Longitude, level indices
```
suggestion implemented

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+     pdel(k) = pint(k+1) - pint(k)
+  end do
+
+  return
```
got rid of it.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+  ! $Id$
+  ! $Author$
+  !
+  !-----------------------------------------------------------------------
```
Removed

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+  character   dim_name*( 256 )
+  integer     var_dimIDs( NF90_MAX_VAR_DIMS )
+  real(r8)    closelat,closelon
+  integer     latidx,lonidx,status,i
```
Updated.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+
+   STATUS = NF90_INQUIRE_VARIABLE( NCID, varID, dimids=var_dimIDs)
+   if ( STATUS /= NF90_NOERR ) then
+      write(iulog,* ) 'ERROR - extractdata.F:Cant get dimension IDs for varid', varid
```
Made changes as recommended.  Also added endrun since this routine assumes that the variable was found and is returning a start count index.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+         usable_var = .true.
+      endif
+   end do
+   return
```
Removed

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+  implicit none
+#if ( defined RS6000 )
+  implicit automatic (a-z)
+#endif
```
Nope

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+    if ( STATUS /= NF90_NOERR ) then
+       STATUS = NF90_INQ_VARID( NCID, 'basedate', bdate_varID )
+       if ( STATUS /= NF90_NOERR )         &
+          write(iulog,*)'ERROR - setiopupdate.F:Cant get variable ID for bdate'
```
Yes. Added endrun calls with message.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+       write(iulog,*) ' Current CAM Date is ',ncdate,' and ',ncsec,' seconds'
+       write(iulog,*) ' IOP start is        ',next_date,' and ',next_sec,'seconds'
+       write(iulog,*) ' IOP end is          ',last_date,' and ',last_sec,'seconds'
+       call endrun
```
done.

jtruesdal commented on src/dynamics/se/advect_tend.F90:
```diff
@@ -85,4 +90,157 @@ subroutine compute_adv_tends_xyz(elem,fvm,nets,nete,qn0,n0)
 
+  !----------------------------------------------------------------------
+  ! computes camiop specific tendencies and
+  ! and writes these to the camiop file
```
done.

jtruesdal commented on src/dynamics/se/advect_tend.F90:
```diff
@@ -85,4 +90,157 @@ subroutine compute_adv_tends_xyz(elem,fvm,nets,nete,qn0,n0)
+      allocate( iop_ttendxyz(nx,nx,nlev,nets:nete) )
+      iop_ttendxyz = 0._r8
+      allocate( iop_ttendxyz_init(nx,nx,nlev,nets:nete) )
+      iop_ttendxyz_init = 0._r8
```
stat and err check done.

jtruesdal commented on src/dynamics/se/advect_tend.F90:
```diff
@@ -85,4 +90,157 @@ subroutine compute_adv_tends_xyz(elem,fvm,nets,nete,qn0,n0)
+      iop_ttendxyz_init = 0._r8
+    endif
+
+    ! save inital/calc tendencies on second call to this routine.
```
fixed

jtruesdal commented on src/dynamics/se/advect_tend.F90:
```diff
@@ -85,4 +90,157 @@ subroutine compute_adv_tends_xyz(elem,fvm,nets,nete,qn0,n0)
+      allocate( out_v(npsq,nlev) )
+      out_v = 0._r8
+      allocate( out_ps(npsq) )
+      out_ps = 0._r8
```
Added stat and endrun checks for all the allocates.

jtruesdal commented on src/dynamics/se/apply_iop_forcing.F90:
```diff
@@ -0,0 +1,241 @@
+  real(r8), intent(in) :: t_in(plev)        ! temperature [K]
+  real(r8), intent(in) :: q_in(plev,pcnst)  ! q tracer array [units vary] already vertically advected
+  real(r8), intent(in) :: t_phys_frc(plev)  ! temperature forcing from physics [K/s]
+  real(r8), intent(in) :: q_phys_frc(plev,pcnst)  ! temperature forcing from physics [K/s]
```
Changed the comment to read ! Change in q due to physics.

jtruesdal commented on src/dynamics/se/apply_iop_forcing.F90:
```diff
@@ -0,0 +1,241 @@
+  ! scm_relaxation is a logical from scamMod
+  ! scm_relax_tau_top_sec and scm_relax_tau_bot_sec are the relaxation times at top and bottom of layer
+  ! also defined in scamMod
+  if ( scm_relaxation.and.scm_relax_linear ) then
```
Removed.  Thanks.

jtruesdal commented on src/dynamics/se/apply_iop_forcing.F90:
```diff
@@ -0,0 +1,241 @@
+     rycept = scm_relax_tau_top_sec - (rslope*scm_relax_top_p)
+  endif
+
+  !    prepare scm_relax_fincl for comparison in scmforecast.F90
```
It exists in the Eul dycore.  Removed the comment.

jtruesdal commented on src/dynamics/se/apply_iop_forcing.F90:
```diff
@@ -0,0 +1,241 @@
+  end do
+
+  do k = 1, plev
+     if( scm_relaxation ) then
```
Right you are.  Removed the if check and reindented the code.

jtruesdal commented on src/infrastructure/phys_grid.F90:
```diff
@@ -203,6 +201,7 @@ subroutine phys_grid_init()
       real(r8),               pointer     :: latvals(:)
       real(r8)                            :: lonmin, latmin
       integer(iMap),          pointer     :: grid_map(:,:)
+      integer(iMap),          pointer     :: grid_map_scm(:,:)
```
The original phys_decomp works just as you described for the EUL dycore, 2d grid for CAM runs or 1d when run as SCAM where both the physics and dynamics are operating on identical grids.  The E3SM SE SCM design (which we are adopting here) has the SE dycore running on a global coarse ne3 grid while the physics is still run in a column and thus requires two grids to handle the physics and dycore.

jtruesdal commented on src/utils/hybvcoord_mod.F90:
```diff
@@ -0,0 +1,28 @@
+module hybvcoord_mod
```
I will need to look into this.

jtruesdal replied:
There is a circular dependency that I don't see an easy way to fix.  I would like to commit this and will continue to look at this in the next set of SCAM mods.

jtruesdal commented on src/dynamics/se/dycore/prim_driver_mod.F90:
```diff
@@ -284,13 +285,20 @@ subroutine prim_run_subcycle(elem, fvm, hybrid,nets,nete, dt, tl, hvcoord,nsubst
+         ! Loop over rsplit vertically lagrangian timesiteps
+         call prim_step_scm(elem, fvm, hybrid,nets,nete, dt, tl, hvcoord,r)
+      else
+         ! Loop over rsplit vertically lagrangian timesiteps
```
Done.

jtruesdal commented on src/dynamics/se/dycore/prim_driver_mod.F90:
```diff
@@ -612,8 +620,81 @@ subroutine prim_step(elem, fvm, hybrid,nets,nete, dt, tl, hvcoord, rstep)
+            hybrid, dt, tl, nets, nete)
+
+       call t_stopf('set_prescribed_scm')
+  enddo
```
Indeed.  Done.

jtruesdal commented on src/physics/cam/check_energy.F90:
```diff
@@ -587,18 +590,34 @@ subroutine check_energy_fix(state, ptend, nstep, eshflx)
     ptend%s(:ncol,:pver) = heat_glob
 
+    if (nstep > 0 .and. write_camiop) then
+      heat_out(:ncol) = heat_glob
+      call outfld('heat_glob',  heat_out(:ncol), pcols, lchnk)
```
This might be a diagnostic that Peter put in.  I'll check with him to see if it can be removed.

jtruesdal replied:
This needs to be redefined as a single global value and not written out as a 2d field.  I will update this.

nusbaume replied:
It looks like this still needs to be updated?

jtruesdal replied:
It's turning out to be more difficult to add this back in as a global and will require pretty extensive changes in addition to restructuring due to a circular dependency.  I would ask to leave it as is for this commit.  It's currently consistent with the way global values were treated by SCAM using the EUL dycore. The global value is written at each lat/lon point and the entire surface field is written to the CAM IOP file. (This is why an ncol array is created and written out). The SCAM physgrid definition only contains 1 column and the appropriate global value is read from the user chosen column.

nusbaume replied:
@jtruesdal Thanks for the explanation!  I am happy to leave this as-is then (at least for now), especially if it is that much effort to implement for arguably minimal gain.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -14,31 +14,46 @@ module scamMod
 use pmgrid,         only: plon, plat, plev, plevp
-use constituents,   only: pcnst
+use constituents,   only: cnst_get_ind, pcnst, cnst_name
+use netcdf,         only: NF90_NOERR,NF90_CLOSE,NF90_GET_VAR,NF90_INQUIRE_DIMENSION, &
```
The vast majority of SCAM has never been updated to use PIO and these changes use the NetCDF interface for consistency.  There is not a good reason to keep using NetCDF although making that happen isn't in the scope of this PR.  Its been on my personal TODO for awhile but I haven't had the time or funding to make it happen yet.

jtruesdal commented on bld/build-namelist:
```diff
@@ -4003,6 +4004,30 @@ if ($dyn eq 'sld') {
+    if ($chem =~ /_mam/) {
+	add_default($nl, 'scm_relax_fincl');
+    }
+    if ($cfg->get('scam_iop')) {
```
Yes.  Done.

jtruesdal commented on cime_config/SystemTests/sct.py:
```diff
@@ -30,15 +30,17 @@ def __init__(self, case):
     def _case_one_setup(self):
         append_to_user_nl_files(caseroot = self._get_caseroot(), component = "cam", contents = "inithist = 'CAMIOP'")
+        if self._case.get_value("CAM_DYCORE") == "se":
+            append_to_user_nl_files(caseroot = self._get_caseroot(), component = "cam", contents = "scale_dry_air_mass     = 0.0D0")
```
I had to put this in specifically to get the tests to pass but can't remember why the defaults were not being applied in this case.  I will double check though.

jtruesdal replied:
It does need to be specified.  Normally for a 3d run the default should be non-zero but in the special case of the bfb check we need to make a 3d run and the companion scam run both with the scale_dry_airmass set to 0.  The defaults to take care of the scam run but the 3d run needs the default overridden for this test.

jtruesdal commented on cime_config/usermods_dirs/scam_camfrc/user_nl_cam:
```diff
@@ -0,0 +1,10 @@
+mfilt=2088
+nhtfrq=1
+co2vmr=368.9e-6
+scm_use_obs_uv = .true.
+scm_relaxation         = .false.
+scm_relax_bot_p              = 105000.
+scm_relax_top_p        = 200.
+scm_relax_linear             = .true.
+scm_relax_tau_bot_sec                = 864000.
+scm_relax_tau_top_sec                = 172800.
```
These lines are here as an example for the user of what a typical set of values for modifying the relaxation algorithm.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -357,9 +385,1244 @@ subroutine scam_readnl(nlfile,single_column_in,scmlat_in,scmlon_in)
+!
+   if ( var_ndims == 0 ) then
+      call endrun('SCAMMOD: var_ndims is 0 for varid:',varid)
+      return
```
Removed.

jtruesdal commented on cime_config/SystemTests/sct.py:
```diff
@@ -47,25 +49,38 @@ def _case_two_setup(self):
+            self._case.set_value("CAM_CONFIG_OPTS","{} -scam camfrc ".format(CAM_CONFIG_OPTS))
+            self._case.set_value("PTS_LAT",44.80320177421346)
+            self._case.set_value("PTS_LON",276.7082039324993)
+            append_to_user_nl_files(caseroot = self._get_caseroot(), component = "cam", contents = "scale_dry_air_mass     = 0.0D0")
```
Same answer as above.

jtruesdal commented:
@cacraigucar I removed the bfb testing line and updated to cam6_4_006 for diagnostic testing.

nusbaume commented on src/dynamics/se/advect_tend.F90:
```diff
@@ -136,19 +139,24 @@ subroutine compute_write_iop_fields(elem,fvm,nets,nete,qn0,n0)
       derivedfq = 0._r8
-      allocate( iop_qtendxyz_init(nx,nx,nlev,pcnst,nets:nete) )
+      allocate( iop_qtendxyz_init(nx,nx,nlev,pcnst,nets:nete),stat=ierr )
+      if (ierr/=0) call endrun( sub//': not able to allocate ipo_qtendxyz' )
```
Typo here (`ipo_qtendxyz` -> `iop_qtendxyz`).

jtruesdal replied:
corrected typo

nusbaume reviewed (changes requested):
Almost there!  Just a few final minor requests and one missed request from last time.

jtruesdal commented on src/control/scamMod.F90:
```diff
@@ -1438,12 +1404,12 @@ subroutine get_start_count (ncid    ,varid  ,scmlat, scmlon, timeidx, start    ,
-  integer     latidx,lonidx,status,i
+  integer            :: dims_set,nlev,var_ndims
+  logical            :: usable_var
+  character(len=256) :: dim_name
```
Added DOI and shr_kind_cl

jtruesdal commented:
It's turning out to be more difficult to write heat_glob as a global and will require pretty extensive changes in addition to restructuring due to a circular dependency. I would ask to leave it as is for this commit. It's currently consistent with the way global values were treated by SCAM using the EUL dycore. The global value is written at each lat/lon point and the entire surface field is written to the CAM IOP file. (This is why an ncol array is created and written out). The SCAM physgrid definition only contains 1 column and the appropriate global value is read from the user chosen column.

nusbaume reviewed (approved):
Everything looks great to me now.  Thanks @jtruesdal!

jtruesdal commented:
> @jtruesdal I tried to run test SCT_D_Ln7.T42_T42_mg17.QPC6.derecho_intel.cam-scm_prep_c6 and SCT_D_P36_Ln7.T42_T42_mg17.QPC6.derecho_intel.cam-scm_prep_c6 and both fail with an mpi error. Can you point me to an SCT test that runs on derecho - or izulme.

@jedwards4b I see that I forgot to address your question about failing SCT tests.  I did correct the error and the tests are running now on both Derecho and Izumi.  The is a baseline difference due to a bug fix for SCAM.  I'm asking for a re-review to clear this issue so I can proceed with the PR.  If you would like to look at the Derecho and Izumi tests they are:
/scratch/cluster/jet/aux_cam_gnu_20240826004033
/scratch/cluster/jet/aux_cam_nag_20240826004050
/glade/derecho/scratch/jet/aux_cam_intel_20240826003035
